### PR TITLE
Add progress state when the drug is being added/updated in `CustomDrugEntrySheet`

### PR DESCRIPTION
**Pull request by janhavisinghh**
https://app.shortcut.com/simpledotorg/story/5045/add-progress-state-when-the-drug-is-being-added-updated

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUpdate.kt [hunk 1]
```diff
@@ -6,7 +6,6 @@ import com.spotify.mobius.Update
 import org.simple.clinic.drugs.PrescribedDrug
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUpdate.kt:6]: The commit message seems to be wrong, can you check it, please?

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryModel.kt [hunk 2]
```diff
@@ -71,4 +78,12 @@ data class CustomDrugEntryModel(
...
     return copy(drugInfoProgressState = IN_PROGRESS)
   }
+
+  fun buttonProgressStateIsSaving(): CustomDrugEntryModel {
+    return copy(saveButtonState = SAVING)
+  }
+
+  fun buttonProgressStateIsSaved(): CustomDrugEntryModel {
+    return copy(saveButtonState = SAVED)
+  }
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryModel.kt:88]: We can just have one function that says, `saveButtonStateChanged` and pass in the params right? 

```kt
fun saveButtonStateChanged(buttonState: ButtonState): CustomDrugEntryModel {
  return copy(saveButtonState = buttonState)
}
```

File: app/src/test/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUpdateTest.kt [hunk 3]
```diff
@@ -107,51 +107,51 @@ class CustomDrugEntryUpdateTest {
...
   }
 
 
   @Test
-  fun `when add button is clicked and the sheet is opened in create mode from drug name with edited dosage and frequency values, then add the drug to the custom drug list`() {
+  fun `when add button is clicked and the sheet is opened in create mode from drug name with edited dosage and frequency values, then add the drug to the custom drug list and set button progress state to saving`() {
     val dosage = "200 mg"
     val frequency = OD
-    val defaultModel = CustomDrugEntryModel.default(openAs = OpenAs.New.FromDrugName(drugName), dosagePlaceholder).drugNameLoaded(drugName)
+    val model = CustomDrugEntryModel.default(openAs = OpenAs.New.FromDrugName(drugName), dosagePlaceholder).drugNameLoaded(drugName).dosageEdited(dosage).frequencyEdited(frequency)
```
Comment (msasikanth) [inline app/src/test/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUpdateTest.kt:131]: Can we break these into multiple lines, please? There is too much happening in a single line.

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUiRenderer.kt [hunk 4]
```diff
@@ -22,6 +22,9 @@ class CustomDrugEntryUiRenderer(
         setSheetTitle(model.drugName, model.dosage, model.drugFrequencyToFrequencyChoiceItemMap[model.frequency]!!.label)
 
       showDefaultDosagePlaceholder(model.dosage, model.dosageHasFocus)
+
+      if (model.isSaveButtonInProgressState)
+        ui.showSaveButtonProgressState()
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryUiRenderer.kt:26]: Just an FYI, if there is no else condition we can just have it in a single line.

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryEffect.kt [hunk 5]
```diff
@@ -43,3 +43,5 @@ data class RemoveDrugFromPrescription(val drugUuid: UUID) : CustomDrugEntryEffec
 object LoadDrugFrequencyChoiceItems : CustomDrugEntryEffect()
 
 object HideKeyboard : CustomDrugEntryEffect()
+
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryEffect.kt:46]: Shouldn't the keyboard changes be in a different PR, considering this PR is for adding progress state in the buttons?

**Review by msasikanth — CHANGES_REQUESTED** [review]
LGTM. Left few comments PTAL

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryModel.kt (hunk 2 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryModel.kt:88]: umm yeah. We can do that in `drugInfoProgressStateLoading` as well right? Can you tell me how do we decide which one is the right approach for doing this in which scenario? I took this approach to maintain consistency in the model for progress states.

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryModel.kt:88]: I guess this is more of a personal preference than the correct approach. For example, if we decided to add another state to the button state, we would need to add another function here. Instead of that, we have been using a single function and just passing the state from the update function. 

You can refer few other places where we used `ButtonState`s.

File: app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryEffect.kt (hunk 5 above)
Comment (janhavisinghh) [reply app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntryEffect.kt]: Raising a different pr for it 👍🏻

**Review by msasikanth — APPROVED** [review]
